src/main/java/backend/Logic.java

```diff
@@ -36,6 +33,7 @@
 
     private RepoIO repoIO;
     public LoginController loginController;
+    private Optional<List<String>> originalLabels = Optional.empty();
```

Comment:
Maybe now isn't the time to implement a fully-fledged undo feature, but we should be very selective of the state we maintain here, because practically every application-level feature will go through this class.

Ideally the state required to undo each action should be encapsulated in some kind of command object. 'Smart actions', basically, which can do and undo themselves.